circom-types/src/groth16/zkey.rs

```diff
@@ -75,246 +48,137 @@ pub struct ZKey<P: Pairing> {
+            s.spawn(|_| b_g2_query = Some(Self::b_g2_query(n_vars, b_g2_section)));
+            s.spawn(|_| l_query = Some(Self::l_query(n_vars - n_public - 1, l_section)));
+            s.spawn(|_| h_query = Some(Self::h_query(domain_size as usize, h_section)));
+        });
 
         let vk = VerifyingKey::<P> {
             alpha_g1: header.verifying_key.alpha_g1,
             beta_g2: header.verifying_key.beta_g2,
             gamma_g2: header.verifying_key.gamma_g2,
             delta_g2: header.verifying_key.delta_g2,
-            gamma_abc_g1: ic,
+            gamma_abc_g1: ic.unwrap()?,
         };
 
         let pk = ProvingKey::<P> {
             vk,
             beta_g1: header.verifying_key.beta_g1,
             delta_g1: header.verifying_key.delta_g1,
-            a_query,
-            b_g1_query,
-            b_g2_query,
-            h_query,
-            l_query,
+            a_query: a_query.unwrap()?,
+            b_g1_query: b_g1_query.unwrap()?,
+            b_g2_query: b_g2_query.unwrap()?,
+            h_query: h_query.unwrap()?,
+            l_query: l_query.unwrap()?,
         };
 
-        Ok(pk)
-    }
-
-    fn get_section(&self, id: u32) -> Section {
-        self.sections.get(&id).unwrap()[0].clone()
-    }
-
-    fn groth_header(&mut self) -> OurResult<HeaderGroth<P>> {
-        let section = self.get_section(2);
-        let header = HeaderGroth::new(&mut self.reader, &section)?;
-        Ok(header)
-    }
-
-    fn ic(&mut self, n_public: usize) -> OurResult<Vec<P::G1Affine>> {
-        // the range is non-inclusive so we do +1 to get all inputs
-        self.g1_section(n_public + 1, 3)
-    }
+        // parse matrices
 
-    /// Returns the [`ConstraintMatrices`] corresponding to the zkey
-    pub fn matrices(&mut self) -> OurResult<ConstraintMatrices<P::ScalarField>> {
-        let header = self.groth_header()?;
+        let mut matrices_section = binfile.take_section(4);
 
-        let section = self.get_section(4);
-        self.reader.seek(SeekFrom::Start(section.position))?;
-        let num_coeffs: u32 = self.reader.read_u32::<LittleEndian>()?;
+        let num_coeffs = u32::deserialize_uncompressed(&mut matrices_section)?;
 
         // instantiate AB
-        let mut matrices = vec![vec![vec![]; header.domain_size as usize]; 2];
+        let mut matrices = vec![vec![vec![]; domain_size as usize]; 2];
         let mut max_constraint_index = 0;
         for _ in 0..num_coeffs {
-            let matrix: u32 = self.reader.read_u32::<LittleEndian>()?;
-            let constraint: u32 = self.reader.read_u32::<LittleEndian>()?;
-            let signal: u32 = self.reader.read_u32::<LittleEndian>()?;
+            let matrix = u32::deserialize_uncompressed(&mut matrices_section)?;
+            let constraint = u32::deserialize_uncompressed(&mut matrices_section)?;
+            let signal = u32::deserialize_uncompressed(&mut matrices_section)?;
```

Comment:
these work because they are little-endian coincidentally, but the CanonicalDeserialize trait always uses little-endian and is non-configurable. Could also use a comment.